File: reference/PCF-DCP-v1.0/src/fragment.rs
//! The Fragment Table: its 9-byte block header and 18-byte entries
//! (spec Section 8).

use crate::consts::{
    ARENA_NONE, FLAG_SHARED, FRAGMENT_ENTRY_SIZE, FRAGTABLE_HEADER_SIZE, KIND_DATA,
};
use crate::error::{Error, Result};

/// One Fragment Entry: a single extent of an inner partition (spec Section 8.2).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FragmentEntry {
    /// Arena-relative start of the extent's bytes.
    pub extent_offset: u64,
    /// Length of the extent in bytes.
    pub extent_length: u64,
    /// Extent kind (`1` = DATA; `0` invalid; `2`/`3` reserved).
    pub kind: u8,
    /// `flags` byte (bit 0 = SHARED; others reserved 0).
    pub flags: u8,
}

impl FragmentEntry {
    /// Serialise to the on-disk 18-byte layout.
    pub fn to_bytes(&self) -> [u8; 18] {
        let mut b = [0u8; 18];
        b[0..8].copy_from_slice(&self.extent_offset.to_le_bytes());
        b[8..16].copy_from_slice(&self.extent_length.to_le_bytes());
        b[16] = self.kind;
        b[17] = self.flags;
        b
    }

    /// Parse from the on-disk 18-byte layout.
    pub fn from_bytes(b: &[u8; 18]) -> Self {
        FragmentEntry {
            extent_offset: u64::from_le_bytes(b[0..8].try_into().unwrap()),
            extent_length: u64::from_le_bytes(b[8..16].try_into().unwrap()),
            kind: b[16],
            flags: b[17],
        }
    }

    /// Whether this entry's `kind` is DATA (the only v1.0 content kind).
    pub fn is_data(&self) -> bool {
        self.kind == KIND_DATA
    }

    /// Whether the SHARED flag (bit 0) is set.
    pub fn is_shared(&self) -> bool {
        self.flags & FLAG_SHARED != 0
    }
}

/// The 9-byte header that begins each Fragment Table block (spec Section 8.1).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FragTableHeader {
    /// Arena-relative offset of the next Fragment Table block of this
    /// partition, or 0 if this is the last block.
    pub next_fragtable_offset: u64,
    /// Number of Fragment Entries packed immediately after this header.
    pub fragment_count: u8,
}

impl FragTableHeader {
    /// Serialise to the on-disk 9-byte layout.
    pub fn to_bytes(&self) -> [u8; 9] {
        let mut b = [0u8; 9];
        b[0..8].copy_from_slice(&self.next_fragtable_offset.to_le_bytes());
        b[8] = self.fragment_count;
        b
    }

    /// Parse from the on-disk 9-byte layout.
    pub fn from_bytes(b: &[u8; 9]) -> Self {
        FragTableHeader {
            next_fragtable_offset: u64::from_le_bytes(b[0..8].try_into().unwrap()),
            fragment_count: b[8],
        }
    }
}

/// Walk an inner partition's Fragment Table chain starting at arena-relative
/// `first_off`, returning its Fragment Entries in logical order across the
/// whole chain (spec Section 8.3). `first_off == 0` yields an empty list.
pub fn walk_fragment_table(arena: &[u8], first_off: u64) -> Result<Vec<FragmentEntry>> {
    let mut out = Vec::new();
    let mut off = first_off;
    // A simple cycle guard: a well-formed chain only ever moves forward, but a
    // corrupt file could loop. Bound the walk by the arena length.
    let mut budget = arena.len() / FRAGTABLE_HEADER_SIZE as usize + 1;
    while off != ARENA_NONE {
        if budget == 0 {
            return Err(Error::OffsetOutOfRange);
        }
        budget -= 1;
        let base = off as usize;
        let hb: [u8; 9] = arena
            .get(base..base + FRAGTABLE_HEADER_SIZE as usize)
            .ok_or(Error::OffsetOutOfRange)?
            .try_into()
            .unwrap();
        let h = FragTableHeader::from_bytes(&hb);
        let mut eo = base + FRAGTABLE_HEADER_SIZE as usize;
        for _ in 0..h.fragment_count {
            let eb: [u8; 18] = arena
                .get(eo..eo + FRAGMENT_ENTRY_SIZE as usize)
                .ok_or(Error::OffsetOutOfRange)?
                .try_into()
                .unwrap();
            out.push(FragmentEntry::from_bytes(&eb));
            eo += FRAGMENT_ENTRY_SIZE as usize;
        }
        off = h.next_fragtable_offset;
    }
    Ok(out)
}

/// Reconstruct the logical content of a partition from its Fragment Entries
/// (spec Section 8.3): concatenate the bytes of its DATA extents in order.
///
/// `arena_used` bounds every extent range; a reserved (non-DATA) kind makes the
/// partition unreadable to a v1.0 reader (spec Section 8.2).
pub fn reconstruct(arena: &[u8], frags: &[FragmentEntry], arena_used: u64) -> Result<Vec<u8>> {
    let mut out = Vec::new();
    for f in frags {
        if !f.is_data() {
            return Err(Error::BadFragmentKind(f.kind));
        }
        let end = f
            .extent_offset
            .checked_add(f.extent_length)
            .ok_or(Error::OffsetOutOfRange)?;
        if end > arena_used || end > arena.len() as u64 {
            return Err(Error::OffsetOutOfRange);
        }
        out.extend_from_slice(&arena[f.extent_offset as usize..end as usize]);
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn entry_roundtrip() {
        let e = FragmentEntry {
            extent_offset: 31,
            extent_length: 6,
            kind: KIND_DATA,
            flags: FLAG_SHARED,
        };
        assert_eq!(FragmentEntry::from_bytes(&e.to_bytes()), e);
        assert!(e.is_data());
        assert!(e.is_shared());
    }

    #[test]
    fn header_roundtrip() {
        let h = FragTableHeader {
            next_fragtable_offset: 0,
            fragment_count: 2,
        };
        assert_eq!(FragTableHeader::from_bytes(&h.to_bytes()), h);
    }
}

File: implementations/ts/pcf-dcp/src/fragment.ts
/**
 * The Fragment Table: its 9-byte block header and 18-byte entries
 * (spec Section 8).
 */

import {
  ARENA_NONE,
  FLAG_SHARED,
  FRAGMENT_ENTRY_SIZE,
  FRAGTABLE_HEADER_SIZE,
  KIND_DATA,
} from "./consts.js";
import { PcfDcpError } from "./errors.js";

/** One Fragment Entry: a single extent of an inner partition (spec 8.2). */
export interface FragmentEntry {
  /** Arena-relative start of the extent's bytes. */
  extentOffset: number;
  /** Length of the extent in bytes. */
  extentLength: number;
  /** Extent kind (`1` = DATA; `0` invalid; `2`/`3` reserved). */
  kind: number;
  /** `flags` byte (bit 0 = SHARED; others reserved 0). */
  flags: number;
}

/** Serialise a Fragment Entry to its on-disk 18-byte layout. */
export function fragmentEntryToBytes(e: FragmentEntry): Uint8Array {
  const b = new Uint8Array(FRAGMENT_ENTRY_SIZE);
  const view = new DataView(b.buffer);
  view.setBigUint64(0, BigInt(e.extentOffset), true);
  view.setBigUint64(8, BigInt(e.extentLength), true);
  b[16] = e.kind & 0xff;
  b[17] = e.flags & 0xff;
  return b;
}

/** Parse a Fragment Entry from its on-disk 18-byte layout. */
export function fragmentEntryFromBytes(b: Uint8Array): FragmentEntry {
  const view = new DataView(b.buffer, b.byteOffset, b.byteLength);
  return {
    extentOffset: Number(view.getBigUint64(0, true)),
    extentLength: Number(view.getBigUint64(8, true)),
    kind: b[16]!,
    flags: b[17]!,
  };
}

/** Whether a Fragment Entry's `kind` is DATA. */
export function isData(e: FragmentEntry): boolean {
  return e.kind === KIND_DATA;
}

/** Whether the SHARED flag (bit 0) is set on a Fragment Entry. */
export function isShared(e: FragmentEntry): boolean {
  return (e.flags & FLAG_SHARED) !== 0;
}

/** The 9-byte header that begins each Fragment Table block (spec 8.1). */
export interface FragTableHeader {
  /** Arena-relative offset of the next block of this partition, or 0. */
  nextFragtableOffset: number;
  /** Number of Fragment Entries packed immediately after this header. */
  fragmentCount: number;
}

/** Serialise a Fragment Table block header to its on-disk 9-byte layout. */
export function fragTableHeaderToBytes(h: FragTableHeader): Uint8Array {
  const b = new Uint8Array(FRAGTABLE_HEADER_SIZE);
  const view = new DataView(b.buffer);
  view.setBigUint64(0, BigInt(h.nextFragtableOffset), true);
  b[8] = h.fragmentCount & 0xff;
  return b;
}

/** Parse a Fragment Table block header from its on-disk 9-byte layout. */
export function fragTableHeaderFromBytes(b: Uint8Array): FragTableHeader {
  const view = new DataView(b.buffer, b.byteOffset, b.byteLength);
  return {
    nextFragtableOffset: Number(view.getBigUint64(0, true)),
    fragmentCount: b[8]!,
  };
}

/**
 * Walk an inner partition's Fragment Table chain starting at arena-relative
 * `firstOff`, returning its Fragment Entries in logical order (spec 8.3).
 */
export function walkFragmentTable(
  arena: Uint8Array,
  firstOff: number,
): FragmentEntry[] {
  const out: FragmentEntry[] = [];
  let off = firstOff;
  let budget = Math.floor(arena.length / FRAGTABLE_HEADER_SIZE) + 1;
  while (off !== ARENA_NONE) {
    if (budget === 0) {
      throw PcfDcpError.offsetOutOfRange();
    }
    budget -= 1;
    if (off + FRAGTABLE_HEADER_SIZE > arena.length) {
      throw PcfDcpError.offsetOutOfRange();
    }
    const h = fragTableHeaderFromBytes(
      arena.subarray(off, off + FRAGTABLE_HEADER_SIZE),
    );
    let eo = off + FRAGTABLE_HEADER_SIZE;
    for (let i = 0; i < h.fragmentCount; i++) {
      if (eo + FRAGMENT_ENTRY_SIZE > arena.length) {
        throw PcfDcpError.offsetOutOfRange();
      }
      out.push(fragmentEntryFromBytes(arena.subarray(eo, eo + FRAGMENT_ENTRY_SIZE)));
      eo += FRAGMENT_ENTRY_SIZE;
    }
    off = h.nextFragtableOffset;
  }
  return out;
}

/**
 * Reconstruct the logical content of a partition from its Fragment Entries
 * (spec Section 8.3): concatenate the bytes of its DATA extents in order.
 */
export function reconstruct(
  arena: Uint8Array,
  frags: readonly FragmentEntry[],
  arenaUsed: number,
): Uint8Array {
  let total = 0;
  for (const f of frags) {
    if (!isData(f)) {
      throw PcfDcpError.badFragmentKind(f.kind);
    }
    const end = f.extentOffset + f.extentLength;
    if (end > arenaUsed || end > arena.length) {
      throw PcfDcpError.offsetOutOfRange();
    }
    total += f.extentLength;
  }
  const out = new Uint8Array(total);
  let p = 0;
  for (const f of frags) {
    out.set(arena.subarray(f.extentOffset, f.extentOffset + f.extentLength), p);
    p += f.extentLength;
  }
  return out;
}
